lib/src/cache/cache.dart

```diff
@@ -0,0 +1,237 @@
+  @mustCallSuper
+  Future<TValue> getAsync(TIdentifier id, Func<Future<TValue>> valueFactory) {
+    _throwWhenDisposed('getAsync');
+
+    // Await any pending cached futures
+    if (_cache.containsKey(id)) {
+      return _cache[id].then((TValue value) async {
+        await onGet(id, value);
+        return value;
+      });
+    }
+
+    // Install Future value
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+    valueFactory.call().then((TValue value) async {
+      await onGet(id, value);
+      _didUpdateController.add(new CacheContext(id, value));
+      completer.complete(value);
+    });
+
+    return _cache[id];
+  }
+
+  /// Is the given identifier in the [Cache]?
+  ///
+  /// If the [Cache] [isDisposedOrDisposing] then a [StateError] is thrown.
+  bool isCached(TIdentifier id) {
+    _throwWhenDisposed('determine if id is cached');
+    return _cache.containsKey(id);
+  }
+
+  /// Updates the current value for a given [TIdentifier] with the given
+  /// [TValue].
+  ///
+  /// Putting a value into the [Cache] results in the imeddiate replacement. any
+  /// future call to [get] or [getAsync] will return the given [value]. Any call
+  /// to put will await the completion of the [onPut] lifecycle method before
+  /// returning the given value. A [CacheContext] event is emitted by the
+  /// [didUpdate] stream upon completion of the [put].
+  ///
+  /// If the [Cache] [isDisposedOrDisposing] then a [StateError] is thrown.
+  Future<TValue> put(TIdentifier id, TValue value) {
+    _throwWhenDisposed('put');
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+
+    onPut(id, value).then((Null _) {
+      _didUpdateController.add(new CacheContext(id, value));
+      completer.complete(value);
+    });
+
+    return _cache[id];
+  }
+
+  /// Allows consumers to define behavior that is executed when a [TValue] is
+  /// obtained from the [Cache].
+  @protected
+  Future<Null> onGet(TIdentifier id, TValue value) async {}
+
+  /// Allows consumers to define behavior that is executed when a [TValue] is
+  /// updated in the [Cache].
+  @protected
+  Future<Null> onPut(TIdentifier id, TValue value) async {}
+
+  /// Allows consumers to define behavior that is executed when a [TValue] is
+  /// removed.
+  @protected
+  Future<Null> onRemove(TIdentifier id, TValue value) async {}
+
+  /// Removes the reference to a [TValue] associated with the given
+  /// [TIdentifier].
+  ///
+  /// If a value is currently cached for the given [TIdentifier], the completion
+  /// of the [get] or [put] that inserted the [TIdentifier] into the cache is
+  /// awaited before removal. This ensures that any pending asyncronous calls
+  /// against the cache resovle to a value.
+  ///
+  /// A [CacheContext] value is emitted by the [didUpdate] stream upon
+  /// successfull removal of the given [TIdentifier].
```

Comment:
The documentation should reflect that the didRemove stream is emitted on when a removal succeeds.